proposals/NNNN-custom-isolation-checking-for-serialexecutor.md

```diff
@@ -0,0 +1,190 @@
+
+The proposal did not account for the situation in which the Swift concurrency runtime has no notion of "current executor" though, causing the following situation to -- perhaps surprisingly -- result in runtime crashes reporting an isolation violation, while in reality, no such violation takes place in the following piece of code:
+
+```swift
+import Dispatch
+
+actor Caplin {
+  let queue: DispatchSerialQueue(label: "CoolQueue")
+  
+  // use the queue as this actor's `SerialExecutor`
+  nonisolated var unownedExecutor: UnownedSerialExecutor {
+    queue.asUnownedSerialExecutor()
+  }
+  
+  nonisolated func connect() {
+    queue.async {
+      // guaranteed to execute on `queue`
+      // which is the same as self's serial executor
+      queue.assertIsolated() // CRASH: Incorrect actor executor assumption
+      self.assertIsolated() // CRASH: Incorrect actor executor assumption
+    }
+  }
+}
+```
+
+One might assume that since we are specifically using the `queue` as this actor's executor... the assertions in the `connect()` function should NOT crash, however how the runtime handles this situation can be simplified to the following steps:
+
+- try to obtain the "current executor"
+- since the current block of code is not executing a swift concurrency task... there is no "current executor" set in the context of `queue.async { ... }`
+- compare current "no executor" to the "expected executor" (the `queue` in our example)
+- crash, as `nil` is not the same executor as the specific `queue`
+
+In other words, these APIs assume to be running "within Swift Concurrency", however there may be situations in which we are running on the exact serial executor, but outside of Swift Concurrency. Isolation-wise, these APIs should still be returning correctly and detecting this situation -- however they are unable to do so, without some form of cooperation with the expected `SerialExecutor`.
+
+## Proposed solution
+
+We propose to add an additional last-resort mechanism to executor comparison, to be used ty the above mentioned APIs.
+
+This will be done by providing a new `checkIsolation()` protocol requirement on `SerialExecutor`:
+
+```swift
+protocol SerialExecutor: Executor {
+  // ...
+ 
+  /// Invoked as last-resort when the swift concurrency runtime is performing an isolation
+  /// assertion, and could not confirm that the current execution context belongs to this
+  /// "expected" executor.
+  ///
+  /// This function MUST crash the program with a fatal error if it is unable 
+  /// to prove that the calling context can be safely assumed to be the same isolation 
+  /// context as represented by this ``SerialExecutor``.
+  ///
+  /// A default implementation is provided that unconditionally causes a fatal error.
+  func checkIsolation()
+}
+
+extension SerialExecutor {
+  public func checkIsolation() {
+    fatalError("Incorrect actor executor assumption, expected: \(self)")
```

Comment:
Can we change this fatal error here since isolation checking is not only for actor executors but can also be done for task executors going forward if they are also `SerialExecutors`.

Reply:
isolation is specifically handled by serial executors; not sure what you mean, should we discuss on the forums with some examples? https://forums.swift.org/t/pitch-custom-isolation-checking-for-serialexecutor/69786/